### DISCUSSION TITLE
Mocking Diesel for unit tests

### DISCUSSION THREAD

**Discussion by naconnors** [Jim-Hodapp-Coaching/ambi-rs#3] [General]
Research discussion to document findings for mocking Diesel ORM to enable unit testing of API endpoints.

Currently, we are not aware of a good way to mock a database connection in Diesel similar to something you would do in unit tests in other languages.

**Comment by naconnors** [comment]
A possible solution would be to use Diesel's [Connection::test_transaction](https://docs.rs/diesel/1.2.0/diesel/connection/trait.Connection.html#method.test_transaction) to prevent changes to the database when running tests.

**Reply by jhodapp** [reply]
The one issue with this solution is to see how Diesel takes in a `Connection` so that it's still possible (and very easy) to run `cargo test` on our repo cloned locally. We should take a deeper look at Diesel's own internal use of `test_transaction`.

**Comment by naconnors** [comment]
Additional possible solution would be to use docker to create a throwaway test database to use and then destroy after running the integration tests.

**Comment by naconnors** [comment]
Additional possible solution would be to use sqlite3 for testing only. We could create the db, run migrations and then cleanup the db once testing is done.
